Add Leith+E

iangrooms opened the pull request:
This commit adds the 2D Leith+E closure, which uses a modified 2D Leith biharmonic viscosity paired with a harmonic backscatter. ('Modified' here is not used in the same sense as 'modified 2D Leith'; it just means that the biharmonic coefficient is modified to account for enstrophy backscatter.) Variables are often named 'leithy' to refer to Leith+E.

The parameterization is controlled by three main entries in user_nl_mom:
1. USE_LEITHY = True
2. LEITH_CK = 1.0
3. LEITH_BI_CONST = 8.0

To use Leith+E you should have LAPLACIAN=True and BIHARMONIC=True. (You should also be explicit and set LEITH_AH=False, along with any other viscous closures. Leith+E does use the background value of the biharmonic coefficient as a minimum, but ignores the background harmonic value.) LEITH_CK is the fraction of energy dissipated by the biharmonic term that gets backscattered by the harmonic term (it's a target; the backscatter rate is not exact.) Recommended values between 0 and 1. LEITH_BI_CONST is Upsilon^6 where Upsilon is the ratio between the grid scale and the dissipation scale for enstrophy. Values should be greater than or equal to 1; 8 is a good place to start.

The code is sensitive to the background value of Ah; specifically, if Ah is too large, the code is unstable. This is because the backscatter coefficient is proportional to Ah, and if Ah is large then you get large backscatter. If your code is unstable, consider reducing, e.g., `AH_VEL_SCALE`.

iangrooms committed:
Add Leith+E

This commit adds the 2D Leith+E closure, which uses a modified 2D Leith biharmonic viscosity paired with a harmonic backscatter. ('Modified' here is not used in the same sense as 'modified 2D Leith'; it just means that the biharmonic coefficient is modified to account for enstrophy backscatter.) Variables are often named 'leithy' to refer to Leith+E.

The parameterization is controlled by three main entries in user_nl_mom:
1. USE_LEITHY = True
2. LEITH_CK = 1.0
3. LEITH_BI_CONST = 8.0

To use Leith+E you should have LAPLACIAN=True and BIHARMONIC=True. (It doesn't hurt to be explicit and also set LEITH_AH=False, along with any other viscous closures, but this is not required. If USE_LEITHY=True it will not use any of the other schemes. It does use the background value of the biharmonic coefficient as a minimum, but ignores the
background harmonic value.) LEITH_CK is the fraction of energy dissipated by the biharmonic term that gets backscattered by the harmonic term (it's a target; the backscatter rate is not exact.) Recommended values between 0 and 1. LEITH_BI_CONST is Upsilon^6 where Upsilon is the ratio between the grid scale and the dissipation scale for enstrophy.
Values should be greater than or equal to 1; 8 is a good place to start.

The code is sensitive to the background value of Ah; specifically, if Ah is too large, the code is unstable. This is because the backscatter coefficient is proportional to Ah, and if Ah is large then you get large backscatter. If your code is unstable, consider reducing, e.g., `AH_VEL_SCALE`.

iangrooms committed:
Background Ah

This commit updates the code so that it uses the background Ah as
a minimum. Previously, if `SMAGORINSKY_AH = True`, Leith+E would
use the Smag value of Ah as the minimum, which is incorrect.

gustavo-marques commented on src/parameterizations/lateral/MOM_hor_visc.F90:
```diff
@@ -1907,6 +2084,10 @@ subroutine hor_visc_init(Time, G, GV, US, param_file, diag, CS, ADp)
+  call get_param(param_file, mdl, "USE_LEITHY", CS%use_Leithy, &
+                 "If true, use a biharmonic Leith nonlinear eddy "//&
+                 "viscosity together with a harmonic backscatter.", &
+                 default=.false., do_not_log=.not.(CS%biharmonic .and. CS%Laplacian))
```
I think `USE_LEITHY` should always be logged and we should add something like "Both BIHARMONIC and LAPLACIAN must be enabled when using this scheme". Otherwise, users will not know if this scheme is available unless `BIHARMONIC = True` and `LAPLACIAN = True`. There is already a check with a FATAL_ERROR to make sure that both of these flags are true when USE_LEITHY is enabled.

iangrooms replied:
Good point. I find myself checking the parameter file to see what is available, so it would be good to have it logged there.

iangrooms committed:
Improve logging

Removed `do_not_log` condition on `USE_LEITHY`

gustavo-marques commented:
Just one more minor very minor suggestion. We should follow [Fortran2003 conventions](https://github.com/NOAA-GFDL/MOM6/wiki/Code-style-guide#mom6-code-style-guide), so
please use <, <=, >, >=, ==, and /= instead of .LT., .LE., .GT., .GE., .EQ., and .NE., respectively.

iangrooms committed:
Fix Leithy Logic

Added one line to fix the fact that the code would only work as
intended if either (i) writing out Ah_h, or (ii) in debug mode.
Also swapped .le. and .lt. for <= and <.

gustavo-marques commented:
@iangrooms, thanks for addressing all the issues.
Our tests are passing (mom_pr). I also ran a few manual tests with this scheme enabled to make sure the model reproduces across restarts and with different PE layouts. Everything checks.
